### ISSUE TITLE
Solution design for deleting historic process resources via REST API

### ISSUE THREAD

**Issue by tmetzke** [camunda/camunda#22850]
# Introduction

The C8 REST API must support the deletion of history data. This includes the following:
 
* Resources like process definitions and all related data, such as finished process instances and their variables
* Specific process instances and all related data, such as their variables

# Concept

* Through C8 REST endpoints, I can do the following:
  * Delete history data related to a single process instance identified by a key
  * Delete history data related to process instances that fulfill certain filter criteria
  * Delete history data related to a single resource identified by a key, such as a process definition (e.g., extend [this endpoint](https://docs.camunda.io/docs/next/apis-tools/camunda-api-rest/specifications/delete-resource/) to also support history deletion optionally or create a new endpoint)
  * Delete history data related to resources, such as process definitions, that fulfill certain filter criteria
* All endpoints must also consider related data, such as all process instances for a process definition and all variables of a process instance.
* Running process instances must have been canceled ([REST API](https://docs.camunda.io/docs/next/apis-tools/camunda-api-rest/specifications/cancel-process-instance/)) or completed (by bringing the instance to a defined end) before being deleted.
* The data is deleted asynchronously in the background through some batching mechanism, so the REST requests return within an appropriate time.
* All endpoints must return a unique identifier of the background (batch) operation that deletes the data. This allows for querying the operation's status with subsequent calls to another dedicated query endpoint.
* The REST gateway issues one command to the brokers containing the data needed to identify and delete the requested data.
* The broker command creates the required background (batch) operation to delete the data in chunks and returns with the unique identifier of that operation.
* The background operation handles the querying and deleting of all related data in chunks asynchronously.
* The background operation is observable on request, yielding its current status (e.g., in progress, failed, completed).

## Deletion "algorithm"

* As we are deleting data chunk by chunk and the data is related to each other, we should follow a clear deletion order that leaves as few unexpected inconsistent data views to users as possible.
* For example, it could make sense to delete "leaf" elements first, e.g., children of a process instance that are unrelated to other data or not parents for other data points themselves, such as variables.
* The deletion strategy could also differ per database type, e.g., deleting leaf entities first by key in RDBMS and deleting by query in ES/OS.
  * For the batch processor, this might change the number of follow-up batch executions to run and the keys contained in there.
  * If all batch executions are established at the beginning, this could be determined upfront by the query results per database type. The query results could differ per database type and thus lead to different batch executions. For example, the RDBMS query implementation would yield the keys of all leaf entities to delete, while the ES/OS implementation returns the process definition or process instance keys only.

# Task Breakdown

- [ ] Design endpoint(s) to delete data asynchronously by query (~1 week)
  - [ ] Define deletion by key for process instance
  - [ ] Define deletion by key for resource, such as process definition
  - [ ] Define deletion by query for process instance
  - [ ] Define deletion by query for resource, such as process definition
- [ ] Provide endpoint(s) to delete data asynchronously by query (~1.5 weeks)
  - [ ] Implement deletion by key endpoint for process instance
  - [ ] Implement deletion by key endpoint for resource, such as process definition
  - [ ] Implement deletion by query endpoint for process instance
  - [ ] Implement deletion by query endpoint for resource, such as process definition
- [ ] Implement querying needed to identify resources to delete (~1.5 weeks)
  - [ ] Query for process instance-related data
  - [ ] Query for resource-related data, e.g., process definition-related data
- [ ] Implement batch processing deletion routines (~4 weeks)
  - [ ] Batch processor deletes history data in chunks from ES/OS
  - [ ] Batch processor deletes history data in chunks from RDBMS

## Out of scope

* If we build this on top of [Batch in Zeebe](https://github.com/camunda/product-hub/issues/2420), all infrastructure of chunked execution and monitoring is out of scope here and handled there

# Discussions and Decisions

The final solution proposal looks as follows:
 
* Deletion request comes in via API, including the query to find resources to delete
* Engine creates a batch operation with the query as context
* Asynchronously, we delete data from the secondary storage via the batch processor
* The exporter is not involved in the deletion to avoid blocking its progress (and thus also engine processing)

![Image](https://github.com/user-attachments/assets/443aa728-1045-41e7-9651-1faa84dfceba)

## Mechanism

* The REST request to delete a process instance or resource by key or query, including history, arrives in the REST gateway.
* The REST gateway sends a command to the broker to create a batch operation for deletion (of a process instance or resources, like a process definition). This includes an abstract search API-level query created from the filter criteria or key provided in the REST request.
* The engine creates a batch operation entity (including the query) and returns its key.
* The command processing commits and returns to the REST API with the key of the batch operation.
* The REST API returns the successful creation of the batch operation, including its reference key, to allow users to track progress.
* The batch processing then happens asynchronously, decoupled from the original request.
* The batch processor queries for all process definitions (or instances) to delete.
* In case of resource deletions, it writes engine commands to remove them from the runtime storage.
* It writes batch executions with all process definitions (or instances) to delete.
* It uses the database client to query for data and to delete data directly. Two possible ways of deleting all related entities:
   1. It deletes all related entities by query, either all data related to the process definition key or broken down by entity
      1. Delete all variables by process definition key
      2. Delete all process instances by process definition key
      3. ...
      4. Delete the process definition by key
   2. It queries for keys of related entities and deletes those by key in chunks of size X
      1. Get X keys of variables related to the process definition key
      2. Delete variables by a list of X keys
      3. Repeat 1. and 2. until the list of keys < X
      4. Run 1., 2., and 3. for the next entity related to the process definition, e.g., the process instance
* Both deletion approaches have pros and cons. Option 1 is less complex for the processor but more likely to time out due to potentially high cardinalities to delete. Option 2 is the opposite: more complex to build but less likely to time out.
* If the keys of entities to delete are gathered upfront (once in the beginning) or while processing (continuous querying) is decided by the [Batch in Zeebe](https://github.com/camunda/product-hub/issues/2420) infrastructure. This topic can handle both options.

## Benefits

* The batch processor takes care of deletions asynchronously, without blocking the engine or exporting
  * **Note**: this assumes that the batch processing is build in such a way, i.e., it does not block the engine, exporting, or other checkers by any tasks it executes.
* The batch processor has full control over how the deletion is chunked, i.e., options 1 and 2 above. There might be more options, e.g., implementing a simple option first and introducing more options as further iterations to provide more resilience to timeouts.

### (1) No multi-region support

* Replication to other regions is currently done by setting up an exporter per region, picking up the events and writing the data accordingly for each region.
* Deleting data via the batch processor lacks awareness for such data layer setups. The batch processor is only aware of the data store attached to it directly, similar to one Operate instance in pre-8.8 architectures.
* As a result, the deletion is currently not replicated to a different region.
  * Note: The batch operation command is distributed to all partitions, so it MIGHT be picked up by a leader operating in another region. However, we can currently not guarantee that leaders exist in all regions, so we cannot guarantee that the batch operation runs in all regions and thus we cannot guarantee that the deletion runs in all regions.

#### Mitigation
 
* We cannot solve this pragmatically since this is an inherent limitation of this solution option
* Product-wise, we are fine with accepting this limitation and need to document this for the feature in our reference docs
* In future versions, we want to achieve multi-region replication on the database level, not the exporter level. This solution option will then become available for multi-region setups as well.

### (2) Concurrent writers on secondary storage

* The exporter is the main writer for the secondary storage and is responsible for all data creation and updates.
* Introducing the batch processor as another writer (doing deletes) can lead to data inconsistencies or run into concurrent modifications with the exporter.
* Additionally, the archiver in ES/OS moves data between indices asynchronously. In RDBMS setups, a history cleanup routine deletes data from tables asynchronously. This can collide with the batch processor, attempting to delete data concurrently with those components.
  * In ES/OS setups, this means that the batch processor might attempt to delete data from indices that have been moved by the archiver already

#### Mitigation

* Concurrency issues between the exporter and archiver/cleanup are managed semantically
  * In ES/OS setups, the exporter only writes into "runtime indices" while the archiver operates on "dated or history indices". The exporter and archiver thus avoid touching the same data concurrently.
  * In RDBMS setups, the exporter writes removal times into data rows that it won't touch again (because their runtime lifecycle is in a final state). The history cleanup only removes rows with removal time. This limits the possibility of touching the same data concurrently.
* Concurrency between the exporter and batch processor can be achieved in the same way
  * In ES/OS setups, the batch processor only queries on "dated/history indices". The batch processor only deletes from "dated/history indices".
  * In RDBMS setups, the batch processor only deletes rows with removal time.
    * Alternatively, the batch processor could also only set removal times to "now" for data rows to be deleted, so the history cleanup cleans them up.
    * This leads to less visibility, though, regarding when the data is actually deleted for search queries, which is unfavorable.
  * This can lead to "leftover" data points
    * ES/OS: Documents that hadn't been in "dated/history indices" when the batch processor queried for keys to delete
    * RDBMS: Rows that didn't have a removal time when the processor deletes the rows.
    * This limitation needs to be documented. In such a case, the user needs to trigger deletion again.
* Concurrency between the batch processor and archiver/cleanup cannot be avoided entirely
  * Both components concurrently operate on history data.
  * The batch processor needs to be implemented in a resilient way with the following scenarios in mind
    * ES/OS: The data to delete can have been found in a dated index that no longer exists because the archiver moved it. The batch processor should track this specifically and consider this data deleted already.
    * RDBMS: The cleanup and batch processor can attempt to delete the same data concurrently. The first transaction to complete will be able to commit; the other will fail and roll back. The cleanup needs to be retried and could yield an empty result - this should be handled identically in the history cleanup and batch processor. The batch processor should track this specifically and consider this data deleted already.
    * RDBMS: The cleanup and batch processor can encounter a deadlock situation where they are competing for resources in a crossover way, where each locks a resource the other one wants to lock as well. This can potentially be limited to very few cases, depending on how the deletion is designed (e.g., delete by query vs. delete by key). Regardless, the cleanup and batch processor can thus encounter deadlock exceptions and need to handle them accordingly, rolling back the transaction and trying again.
      * Alternatively, the batch processor could ignore rows with a removal time lower than or equal to now because the history cleanup will delete them in the next run anyway.
      * However, this can lead to a time window in which the batch processor considers data deleted that is, in fact, not deleted until the cleanup runs again. The batch processor is unaware of the next cleanup run, though.

# Links

- Product Hub (internal): https://github.com/camunda/product-hub/issues/2472

**Comment by tmetzke** [comment]
Overall concept idea for deleting resources (might have to go into a separate epic) and process instances:


### **Delete Resource**

**Status quo**: Zeebe delete resource API will only delete definition in  
Zeebe/Runtime, prerequisite only if all instances have finished  
Operate handles deletion of historic data autonomously + deletes definition via Zeebe API from runtime.

**Problem:**  
With migration to V2 API, we need to migrate historic data cleanup. It doesn’t fit the REST API service scope as it requires aonger running batch delete job on db level.

**Solution Design considerations for changing the Delete Resource API implementation:**  
Shall the API delete history as well => **decision: not by default**  
This would prevent the use case to only delete the definition to prevent new instances of a version/definition.

Shall it delete transitively by flag? (history) => preferred  
Edge case to consider: delete in operate when it already got deleted in Zeebe

Design of transitive delete => challenge: potential high cardinality of instances to delete, does not fit functionality of exporter as it requires a stateful batch delete job.  
Idea: exporter only deletes definition, new archiver cares about cleaning up instances for which the definition is gone

Needs handling of edge-case where the definition is already gone from runtime state  
Idea: add dedicated flag to not fail in such a case

General Zeebe behavior on resource delete does not change, except handling the do not fail if not present flag. Resource delete event record needs to get extended with lag for transitive/historic data delete.

### **Delete Instance:**

**Status quo:** we have cancel only on Zeebe, in Operate you could delete finished instances individually

**Solution Design considerations for a Delete Instance API implementation:**  
Zeebe as event bus => record with no functional impact in engine => rejected if instance exists in runtime still.  
Exporter deleted instance by ID on deleted instance record.

Nice to have: flag to implicitly cancel a potentially running instance

**Comment by ralfpuchert** [comment]
## Endpoint

`DELETE /v2/process-instances/{processInstanceKey}` sends delete command to Zeebe.
Returns whether the command was successfully sent. 
There is no immediate response whether the process instance was deleted. 
Only that the deletion was requested.

## Zeebe command

Zeebe issues a `ProcessInstance.DELETED` event. 
This happens also if the process instance was not deleted via REST endpoint.

## Exporter

1. Receives `ProcessInstance.DELETED` event
2. Retrieves process instance key
3. Uses [ProcessInstanceWriter::deleteInstanceById](https://github.com/camunda/camunda/blob/f9a67af561e9ba0b5cef6af069cb4791a3cf9902/operate/webapp/src/main/java/io/camunda/operate/webapp/writer/ProcessInstanceWriter.java#L18-L18) to delete all data associated with process instance. ([ES implementation](https://github.com/camunda/camunda/blob/f9a67af561e9ba0b5cef6af069cb4791a3cf9902/operate/webapp/src/main/java/io/camunda/operate/webapp/elasticsearch/writer/ProcessInstanceWriter.java#L58-L58))

This happens asynchronously.
Question: What happens in the moment partly data were deleted and user requests the process instance data? 

It does the following:
1. Check if process instance is deletable 
   * `State` needs to be `COMPLETED` or `CANCELED`
   * `endDate` must be set and before now
2. Each dependant data (in corresponding Operate indices) will be deleted:
   * Decision instance
   * Event
   * Flow node instance
   * Incident
   * Job
   * Post importer
   * Sequence flow
   * User task
   * variable
3. Process instance itself will be deleted

**Comment by tmetzke** [comment]
As indicated in https://github.com/camunda/product-hub/issues/2472, such deletion features are mostly meant for test processes and exist in addition to the ILM policy. What is true for process and decision definitions is also true for process instance deletions.

As such, we can also go for a limited solution for 8.7 and implement and document a limited cardinality with which this works.
We have the same thing in C7 where synchronous cascading deletes only work up to a certain cardinality.

**Comment by tmetzke** [comment]
## Decision: postponed to 8.8

* This feature will not be built with 8.7 due to team capacity constraints.
* The Operate UI is not affected and will remain functioning as is.
* The C8 API will only allow deleting from runtime with 8.7. No history deletion can directly be triggered from the C8 API with 8.7.

With 8.8, we aim to provide a consistent history deletion experience via API, implementing the abovementioned tasks.

**Comment by romansmirnov** [comment]
> Idea: exporter only deletes definition, new archiver cares about cleaning up instances for which the definition is gone

@tmetzke, how does it determine which definitions got deleted? Does it determine the deleted definitions dynamically by periodically doing a query to get all definitions (that do not exist but are referenced)? Or would it mean there is some state that keeps track of deleted definitions? If there is one, it will delete all process definition-related data (instances, ...) and only then remove that deleted definition from that state. Alternatively, it could mark the definition as deleted and eventually delete it once all related data was deleted.

What do you think about RDBMS? Would we do the same there? With RDBMS, we would need a mechanism to ensure the transaction does not fail because it gets too big or timed out, eventually ending in a loop that cannot be recovered from.

Looking at it also from cross-cutting features like backup & restore. An upcoming initiative is to decouple the backup between the primary storage (C8/Zeebe) and the secondary storage (ES/OS/RDBMS) to simplify the backup procedures, etc. In that context, specifically looking at ES/OS, it includes that we will no longer create an ES snapshot of indices in a specific order. The challenge is that ES snapshots are [not point-in-time](https://www.elastic.co/guide/en/elasticsearch/reference/current/snapshot-restore.html#snapshot-start-stop-times). Meaning that their snapshot has a start and end time, and the snapshot represents a view of each index's data at some point between these two times.

Bringing this together, while ES takes a snapshot, a user requests to delete the definition with the cascading/transitive property enabled. The ES snapshot might result in a state still containing the (deleted) definition. Additionally, some process instances related to that definition might have already been deleted.
Now, the question is, when restoring from that ES snapshot, can we replay everything to ensure that the definition gets deleted eventually, and thus, all process instances related to it get deleted as well?
* As long as we try to replay it from the secondary storage, this won't work.
* It must be replayed (re-exporting the definition deletion event) from the primary storage (C8/Zeebe). But how do we detect that case?
  * It would work if the primary storage was restored from a point before the deletion event so that it gets replayed and re-exported.
  * If it restores from a later point (after the deletion event), it won't recognize that.

The best solution would be if the primary storage had a task queue that the camunda exporter (ES/OS and RDBMS) could use to execute the task-related operations on the secondary storage. The primary storage ensures the state of the task queue is correct by applying state changes as with any other data stored in the primary storage. Additionally, the camunda exporter has a way to notify the primary storage that the task is done, which can also be replayed.

Hmm... that would explode this issue 😄 But I believe we need something similar in other cases as well. I would just avoid doing double work.

Okay, but let's go one step back again:
* Assuming the aim is 8.8 to release this feature described in this issue. Then, the old backup & restore procedure applies.
* That backup procedure results in a strong, consistent backup across the primary and secondary storage. Eventually, that setup will work, depending on some implementation details.
* If the backup & restore procedure gets re-designed, that approach would potentially not work anymore, and we would need to refactor it.

Should we sync on this topic?

**Comment by tmetzke** [comment]
Thank you, @romansmirnov, for those data store-related insights and thoughts. That is precisely what I was lacking in understanding so far 👍 

Syncing on this sounds reasonable to me. I'll set up a time for us to do an initial brainstorming, and then I'll refine and document the adjusted solution design.

**Comment by tmetzke** [comment]
# Update on solution design

@romansmirnov and I met today to discuss this. The results are on this (🔒 internal) [Miro board](https://miro.com/app/board/uXjVIYWBnio=/).

## Current idea

This is a high-level sequence flow for process definition deletion:
![Image](https://github.com/user-attachments/assets/350368aa-b497-4e31-bdbe-3f15ce0a81b0)

The basics:
* The REST request to delete a definition by key, including history, arrives in the REST gateway.
* The REST gateway sends a command to the broker (like it already does), including the flag to cascade to history.
* The engine deletes the process definition from runtime (like it already does).
* The engine creates a batch operation, providing the action DELETE and the process definition key as context.
* The batch processing creates a batch context, generates a key for that, and returns that key.
* The command processing commits and returns to the REST API, through the engine, as usual.
* The REST API returns the successful deletion from runtime and the batch operation reference key to allow tracking secondary storage progress.
* The batch processing happens asynchronously, decoupled from the original request, fetching the process instances to delete and deleting them (including all attached child elements like variables) in chunks.

### Benefits

* This solves the "[what happens on backup & replay](https://github.com/camunda/camunda/issues/22850#issuecomment-2682372351)" issue
  * The batch processor keeps a task queue that is resilient to and consistent across backups and replays. We replay from the primary storage (assuming the batch processor operates using commands and events). The batch processor brings along the infrastructure to determine when it's done.
* Batch data responsibilities can be clearly defined
  * Single deletion use case (delete one resource, e.g. one definition or one instance)
    * The REST gateway passes the process definition key to the engine and creates a DELETE command with it
    * The engine creates a batch operation for deleting resources and passes on the process definition key
    * The batch processor contains logic to find the process instances (and other child resources) to delete (happens in chunks already, ideally, because it can be many)
    * The batch processor contains logic to delete process instances in chunks
  * Delete by query use case
    * The REST gateway receives a user-defined query to delete by, e.g., "delete all process definitions with processDefinitionId = 'my-process' "
    * The REST gateway creates a Search API-level query from the REST DTO
    * The REST gateway passes the query to the engine with the DELETE command
    * The engine creates a batch operation for deleting resources and passes on the query
    * The batch processor contains logic to find the process instances (and other child resources) to delete based on the query (happens in chunks already, ideally, because it can be many)
    * The batch processor contains logic to delete process instances in chunks

### Requirements from the batch processing

* The engine creates the batch operation request transparently
  * Commands stay atomic, when spanning over primary and secondary storage, i.e. deleting a process definition from runtime and history
  * Callers get a response that represents the whole operation, i.e. deleting from runtime and creating the batch operation to delete from history

### Alternative

Assuming the engine does not transparently create the batch operation, i.e. the batch processing needs to be invoked on gateway service level separately, the following would be necessary:
 
* We receive a request to delete a definition by key, including all related history
* We create an engine command to delete from runtime
* The runtime command succeeds
* We have to use the gateway service for initiating a batch operation
  * The main issue here is we're missing atomicity for this action (deleting the definition in runtime and history).
  * The runtime changes are committed at this point. The batch operation for deleting from history is not yet, though.
  * This invocation can fail for any reason (e.g. secondary storage is not accessible, batch operation cannot be created in the engine)
  * We need to handle that the runtime deletion is done and committed but the batch operation creation failed.
    * Do we issue a failed status overall? The runtime deletion is committed, though, so it was partially successful. Can a user simply retry the REST command?
  * We're lacking traceability and replay-ability of the overall action of deleting the definition since we have no idea of the need to create a batch operation in the engine as well yet. The first engine command was only concerned with deleting from runtime.

## Next steps

* Sync with @mustafadagher and team to assess if our vision of how we can use batch processing is feasible.

**Comment by tmetzke** [comment]
# Update#2 on solution design

@mustafadagher and I met today to discuss a potential solution design, considering the latest results of the "Batch in Zeebe" POC.

## Updated idea

This is a refined high-level sequence flow for resource deletion:
![Image](https://github.com/user-attachments/assets/6d7200e5-0148-4cec-a274-98ea19310e37)

The basics:
* The REST request to delete resources (by query or by key), including history, arrives in the REST gateway.
* _Changed_: The REST gateway sends a command to the broker to delete a resource asynchronously
* The engine creates a batch operation with a key, providing the action DELETE_PROCESS_DEFINITION and the user query on search API-level as context.
* The command processing commits and returns to the REST API, through the engine, as usual.
* The REST API returns the successful deletion from runtime and the batch operation reference key to allow tracking secondary storage progress.
* The batch processing happens asynchronously, decoupled from the original request, fetching the process instances to delete and deleting them (including all attached child elements like variables) in chunks.
* _New_: The batch processor has logic to derive what it needs to find the definitions based on the query, the related process instances, variables, etc.
* _New_: The batch processor writes the necessary DELETE commands to the engine log
* _New_: The engine creates DELETE events and deletes from runtime (if still there)
* _New_: The exporters pick up the DELETE events and delete from secondary storage
* _New_: The batch processor triggers follow-up actions until the deletion is completed
 
## Benefits
 
* If we build on top of Batch in Zeebe like this, the "History Deletion" team only needs to
  * Implement missing DELETE intents in the brokers
  * Implement required querying routines into the batch processor for deletion of process definitions, process instances, and whatever else needs to be supported
  * Implement the REST endpoint(s) needed for this
    * _Note_: this could also be handled by the Operate UI migration team, depending on capacity
  * **Highest risk**: Investigate how deletion from secondary storage can be picked up and done by exporters

## Action items
* Sync on the initial REST API design in late March (week of 24th March) with Operate UI migration, API, and Batch in Zeebe teams
* Break down the remaining tasks that are exclusive to the History Deletion topic

**Comment by romansmirnov** [comment]
> Highest risk: Investigate how deletion from secondary storage can be picked up and done by exporters

I would be interested to iterate on the options you may consider for solving this.

**Comment by tmetzke** [comment]
# Update#3 on solution design

@romansmirnov and I synced today about the possible solutions for data deletion from the secondary storage, which was identified as the highest risk of the current design.

**TL;DR**: There is no optimal solution for this. All options we know of come with certain drawbacks, either in complexity or supported C8 setups.

## Options

We discussed the following options. Each option describes how it works, where the risks lie, and how to cope with them.

### Deletion via batch processor

This refers roughly to [solution proposal no.1](https://github.com/camunda/camunda/issues/22850#issuecomment-2690576880). Overall, the batch processor is responsible for deleting data from the secondary storage.

#### Mechanism

* The batch processor receives a set of keys or a query for process definitions to delete.
* It queries for all process definitions to delete.
* It writes an engine command ProcessDefinition.DELETE for all process definition keys to remove the process from the runtime storage.
* It writes a batch execution with all process definition keys.
* It uses the database client to query for data and to delete data directly. Two possible ways of deleting all related entities:
   1. It deletes all related entities by query
      1. Delete all variables by process definition key
      2. Delete all process instances by process definition key
      3. ...
   2. It queries for keys of related entities and deletes those by key in chunks of size X
      1. Get X keys of variables related to the process definition key
      2. Delete variables by a list of X keys
      3. Repeat 1. and 2. until the list of keys < X
      4. Run 1., 2., and 3. for the next entity related to the process definition, e.g., the process instance
* Both deletion approaches have pros and cons. Option 1 is less complex for the processor but more likely to time out due to potentially high cardinalities to delete. Option 2 is the opposite: more complex to build but less likely to time out.

#### Benefits

* The batch processor takes care of deletions asynchronously, without blocking the engine or exporting
  * **Note**: this assumes that the batch processing is build in such a way, i.e., it does not block the engine, exporting, or other checkers by any tasks it executes. This is currently under investigation/in development.
* The batch processor has full control over how deletion is chunked, i.e., options 1 and 2 above. There might be even more options and even fallbacks if a simple one times out and needs to be chunked down further.

#### Risks

* Not suitable for multi-region
  * Deleting from the batch processor lacks awareness for more complex data layer setups.
  * The batch processor is only aware of the data store attached to it directly, similar to one Operate instance in pre-8.8 architectures.
  * As a result, the deletion is not replicated to a different region, for example. The batch processor executing the batch operation operates in one region only.
    * **Note**: The batch operation command is distributed to all partitions, so it MIGHT be picked up by a leader operating in another region. However, we can currently not guarantee that leaders exist in all regions, so we cannot guarantee that the batch operation runs in all regions and thus we cannot guarantee that the deletion runs in all regions.

### Deletion via exporter

This refers roughly to [solution proposal no.2](https://github.com/camunda/camunda/issues/22850#issuecomment-2713726576). The batch processor writes DELETE commands to the engine log. The exporter is responsible for deleting data from the secondary storage.

#### Mechanism

* The batch processor receives a set of keys or a query for process definitions to delete.
* It queries for all process definitions to delete. It queries for all related entities' keys to delete.
* It writes an engine command with DELETE intent for all related entities (to be decided if multiple commands or one commands and how many keys per command).
* The engine processes the command(s) and write DELETED events accordingly (might need new events emitted by the engine).
* _to be continued..._

#### Benefits

* 

#### Risks

*

**Comment by tmetzke** [comment]
# Further input for solution design

@ChrisKujawa, @RomanJRW, and I synced today on the current solution options to dig deeper into possible deletion options from secondary storage.

## Open questions/topics

* @ChrisKujawa: data inconsistency is something to really watch out for here since we're introducing a second writer next to the exporter
  * With only history data (instances need to be completed/canceled before they can be deleted) it is less risky but not completely gone
  * Can we guarantee we always only support deleting history data (finished/canceled instances)? If not, we might run into concurrency issues at some point without being aware.
* @ChrisKujawa: Can we reuse the archiver mechanism somehow?
  * Create a marker with a key list of process definitions to delete through the exporter
  * This then takes the usual route of writing data but is not as write-heavy as a deletion
  * This is picked up by the archiver, moving the related data "out of sight" (e.g. by dropping it)
  * @RomanJRW: Optionally, until the data is removed, we could maybe even consider the list of marked keys in search requests and treat the data as deleted

## Decisions

* We should spent some time in a PoC to evaluate the marker & async deletion option

## Action items
* @RomanJRW will talk to Nicolas P.P. regarding potential capacity for such a PoC (2 engineers, e.g., @tmetzke and an engineer with more data layer experience)

**Comment by tmetzke** [comment]
# Considerations for deletion via batch processor

@romansmirnov and I synced up on this solution option again to understand the potential risks better and learn how to mitigate them.

## No multi-region support

### Risk

* Replication to other regions is currently done by setting up an exporter per region, picking up the events and writing the data accordingly for each region.
* Deleting data via the batch processor lacks awareness for such data layer setups. The batch processor is only aware of the data store attached to it directly, similar to one Operate instance in pre-8.8 architectures.
* As a result, the deletion is currently not replicated to a different region.
  * Note: The batch operation command is distributed to all partitions, so it MIGHT be picked up by a leader operating in another region. However, we can currently not guarantee that leaders exist in all regions, so we cannot guarantee that the batch operation runs in all regions and thus we cannot guarantee that the deletion runs in all regions.

### Mitigation
 
* We cannot solve this pragmatically since this is an inherent limitation of this solution option
* Product-wise, we are fine with accepting this limitation and need to document this for the feature in our reference docs
* In future versions, we want to achieve multi-region replication on the database level, not the exporter level. This solution option will then become available for multi-region setups as well.

## Concurrent writers on secondary storage

### Risk

* The exporter is the main writer for the secondary storage and is responsible for all data creation and updates.
* Introducing the batch processor as another writer (doing deletes) can lead to data inconsistencies or run into concurrent modifications with the exporter.
* Additionally, the archiver in ES/OS moves data between indices asynchronously. In RDBMS setups, a history cleanup routine deletes data from tables asynchronously. This can collide with the batch processor, attempting to delete data concurrently with those components.
  * In ES/OS setups, this means that the batch processor might attempt to delete data from indices that have been moved by the archiver already

### Mitigation

* Concurrency issues between the exporter and archiver/cleanup are managed semantically
  * In ES/OS setups, the exporter only writes into "runtime indices" while the archiver operates on "dated or history indices". The exporter and archiver thus avoid touching the same data concurrently.
  * In RDBMS setups, the exporter writes removal times into data rows that it won't touch again (because their runtime lifecycle is in a final state). The history cleanup only removes rows with removal time. This limits the possibility of touching the same data concurrently.
* Concurrency between the exporter and batch processor can be achieved in the same way
  * In ES/OS setups, the batch processor only queries on "dated/history indices". The batch processor only deletes from "dated/history indices".
  * In RDBMS setups, the batch processor only deletes rows with removal time.
    * Alternatively, the batch processor could also only set removal times to "now" for data rows to be deleted, so the history cleanup cleans them up.
    * This leads to less visibility, though, regarding when the data is actually deleted for search queries, which is unfavorable.
  * This can lead to "leftover" data points
    * ES/OS: Documents that hadn't been in "dated/history indices" when the batch processor queried for keys to delete
    * RDBMS: Rows that didn't have a removal time when the processor deletes the rows.
    * This limitation needs to be documented. In such a case, the user needs to trigger deletion again.
* Concurrency between the batch processor and archiver/cleanup cannot be avoided entirely
  * Both components concurrently operate on history data.
  * The batch processor needs to be implemented in a resilient way with the following scenarios in mind
    * ES/OS: The data to delete can have been found in a dated index that no longer exists because the archiver moved it. The batch processor should track this specifically and consider this data deleted already.
    * RDBMS: The cleanup and batch processor can attempt to delete the same data concurrently. The first transaction to complete will be able to commit; the other will fail and roll back. The cleanup needs to be retried and could yield an empty result - this should be handled identically in the history cleanup and batch processor. The batch processor should track this specifically and consider this data deleted already.
    * RDBMS: The cleanup and batch processor can encounter a deadlock situation where they are competing for resources in a crossover way, where each locks a resource the other one wants to lock as well. This can potentially be limited to very few cases, depending on how the deletion is designed (e.g., delete by query vs. delete by key). Regardless, the cleanup and batch processor can thus encounter deadlock exceptions and need to handle them accordingly, rolling back the transaction and trying again.
      * Alternatively, the batch processor could ignore rows with a removal time lower than or equal to now because the history cleanup will delete them in the next run anyway.
      * However, this can lead to a time window in which the batch processor considers data deleted that is, in fact, not deleted until the cleanup runs again. The batch processor is unaware of the next cleanup run, though.

**Comment by tmetzke** [comment]
I have updated the issue description with the solution proposal and task breakdown.
All issue comments related to this solution finding have been marked as resolved to remove clutter from the issue.
The [considerations for deleting from the exporter](https://github.com/camunda/camunda/issues/22850#issuecomment-2736834299) is an alternative approach to the proposed solution and thus remains visible in the comment streams.

**Comment by tmetzke** [comment]
Closing this as done, the issue description contains the final solution proposal.
